src/video.rs

```diff
@@ -105,68 +118,65 @@ pub fn compare_videos(
-                let dst_frame = distorted.read_video_frame::<u8>();
-                if let (Some(src_frame), Some(dst_frame)) = (src_frame, dst_frame) {
-                    let src = Yuv::new(src_frame, src_config).unwrap();
-                    let dst = Yuv::new(dst_frame, dst_config).unwrap();
-                    (Rgb::try_from(src).unwrap(), Rgb::try_from(dst).unwrap())
-                } else {
-                    break;
-                }
-            }
-            (_, _) => {
-                let src_frame = source.read_video_frame::<u16>();
-                let dst_frame = distorted.read_video_frame::<u16>();
-                if let (Some(src_frame), Some(dst_frame)) = (src_frame, dst_frame) {
-                    let src = Yuv::new(src_frame, src_config).unwrap();
-                    let dst = Yuv::new(dst_frame, dst_config).unwrap();
-                    (Rgb::try_from(src).unwrap(), Rgb::try_from(dst).unwrap())
+    let (result_tx, result_rx) = mpsc::channel();
+    let src_bd = source.get_bit_depth();
+    let dst_bd = distorted.get_bit_depth();
+
+    let decoders = Arc::new(Mutex::new((source, distorted)));
+    for _ in 0..threads {
+        let decoders = Arc::clone(&decoders);
+        let result_tx = result_tx.clone();
+
+        std::thread::spawn(move || {
+            loop {
+                let score = match (src_bd, dst_bd) {
+                    (8, 8) => calc_score::<u8, u8>(&decoders, &src_config, &dst_config),
+                    (8, _) => calc_score::<u8, u16>(&decoders, &src_config, &dst_config),
+                    (_, 8) => calc_score::<u16, u8>(&decoders, &src_config, &dst_config),
+                    (_, _) => calc_score::<u16, u16>(&decoders, &src_config, &dst_config),
+                };
+
+                if let Some(result) = score {
+                    result_tx.send(result).unwrap();
                 } else {
+                    // no score = no more frames to read
                     break;
                 }
             }
-        };
-        let src_xyb = Xyb::try_from(src_rgb).unwrap();
-        let dst_xyb = Xyb::try_from(dst_rgb).unwrap();
-        let result =
-            compute_frame_ssimulacra2(src_xyb, dst_xyb).expect("Failed to calculate ssimulacra2");
-        if verbose {
-            println!("Frame {frame}: {result:.8}");
-        }
-        results.push(result);
+        });
+    }
+
+    // Needs to be dropped or the main thread never stops waiting for scores
+    drop(result_tx);
+
+    let mut results = Vec::new();
+    let mut frame = 0usize;
+
+    let progress = if stderr().is_tty() {
+        ProgressBar::new_spinner().with_style(
+            ProgressStyle::with_template(
+                "[{elapsed_precise:.blue}] [{per_sec:.green}] Frame {pos}",
+            )
+            .unwrap()
+            .with_key("per_sec", |s: &ProgressState, w: &mut dyn std::fmt::Write|
+                write!(w, "{:5.02} FPS", s.per_sec()).unwrap()
```

Comment:
Nit: I personally prefer lowercase "fps" but I don't really care that much in the end.